docs/architecture/psa-driver-interface.md

```diff
@@ -0,0 +1,561 @@
+
+The driver interface allows implementations of the PSA Crypytography API to be built compositionally. An implementation of the PSA Cryptography API is composed of a **core** and zero or more **drivers**. The core handles key management, enforces key usage policies, and dispatches cryptographic operations either to the applicable driver or to built-in code.
+
+Functions in the PSA Cryptography API invoke functions in the core. Code from the core calls drivers as described in the present document.
+
+### Types of drivers
+
+The PSA Cryptography driver interface supports two types of cryptoprocessors, and accordingly two types of drivers.
+
+* **Transparent** drivers implement cryptographic operations on keys that are provided in cleartext at the beginning of each operation. They are typically used for hardware **accelerators**. When a transparent driver is available for a particular combination of parameters (cryptographic algorithm, key type and size, etc.), it is used instead of the default software implementation. Transparent drivers can also be pure software implementations that are distributed as plug-ins to a PSA Crypto implementation.
+* **Opaque** drivers implement cryptographic operations on keys that can only be used inside a protected environment such as a **secure element**, a hardware security module, a smartcard, a secure enclave, etc. An opaque driver is invoked for the specific key location that the driver is registered for: the dispatch is based on the key's lifetime.
+
+### Requirements
+
+The present specification was designed to fulfil the following high-level requirements.
+
+[Req.plugins] It is possible to combine multiple drivers from different providers into the same implementation, without any prior arrangement other than choosing certain names and values from disjoint namespaces.
+
+[Req.compile] It is possible to compile the code of each driver and of the core separately, and link them together. A small amount of glue code may need to be compiled once the list of drivers is available.
+
+[Req.types] Support drivers for the following types of hardware: accelerators that operate on keys in cleartext; cryptoprocessors that can wrap keys with a built-in keys but not store user keys; and cryptoprocessors that store key material.
+
+[Req.portable] The interface between drivers and the core does not involve any platform-specific consideration. Driver calls are simple C functions. Interactions between driver code and hardware happen inside the driver (and in fact a driver need not involve any hardware at all).
+
+[Req.location] Applications can tell which location values correspond to which secure element drivers.
```

Comment:
'Can tell' is ambiguous here. Does it mean that an application knows which driver is attached to a certain location value, or does it mean the application can specify which driver needs to be attached to a given location value?

Reply:
I'd expect the location value to be hard-coded in core and driver binaries. Adding an indirection layer to renumber locations per application would be a significant cost on a microcontroller scale and I'm not seeing a need for it. I expect a header with `#define ACME_DRIVER_KEY_LOCATION 42` to be part of the output of the driver integration process (clarifying this is part of the TODO in the section “Using opaque drivers”).

Reply:
So the intent here is that each driver defines their own location value? Then the requirement language probably needs to be updated accordingly.

Reply:
I think the requirement text is sufficient, but I've added a concrete mechanism: https://github.com/ARMmbed/mbedtls/pull/3493/commits/71db60bd11a0ccd09a2a5e1222788150846c5a86